checker/src/org/checkerframework/checker/regex/RegexQualifiedTransfer.java

```diff
@@ -45,82 +59,218 @@ public RegexQualifiedTransfer(QualAnalysis<QualParams<Regex>> analysis) {
+        res = super.visitGreaterThan(n, in);
+        res = handleMatcherGroupCount(n.getLeftOperand(), n.getRightOperand(), false, in, res);
+        return res;
+    }
+
+    @Override
+    public
+    TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>>
+    visitGreaterThanOrEqual(GreaterThanOrEqualNode n,
+            TransferInput<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> in) {
+        // Look for: mat.groupCount() >= constant
+        // Make mat be @Regex(constant)
+        TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> res;
+        res = super.visitGreaterThanOrEqual(n, in); 
+        res = handleMatcherGroupCount(n.getLeftOperand(), n.getRightOperand(), true, in, res);
+        return res;
+    }
+
+    /**
+     * See whether possibleMatcher is a call of groupCount on a Matcher and
+     * possibleConstant is a constant.
+     * If so, annotate the matcher as
+     * constant + 1 if !isAlsoEqual
+     * constant if isAlsoEqual
+     *
+     * @param possibleMatcher the Node that might be a call of Matcher.groupCount()
+     * @param possibleConstant the Node that might be a constant
+     * @param isAlsoEqual whether the comparison operation is strict or reflexive
+     * @param in the input TransferResult
+     * @return the possibly refined output TransferResult
+     */
+    private
+    TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>>
+    handleMatcherGroupCount(
+            Node possibleMatcher, Node possibleConstant, boolean isAlsoEqual,
+            TransferInput<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> in,
+            TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> resultIn) {
+        if (!(possibleMatcher instanceof MethodInvocationNode)) {
+            return resultIn;
+        }
+        if (!(possibleConstant instanceof IntegerLiteralNode)) {
+            return resultIn;
+        }
+
+        MethodInvocationNode min = (MethodInvocationNode) possibleMatcher;
+        MethodAccessNode target = min.getTarget();
+        ExecutableElement method = target.getMethod();
+
+        if (!ElementUtils.matchesElement(method, GROUP_COUNT_METHOD_NAME)) {
+            return resultIn;
+        }
+
+        Node receiver = target.getReceiver();
+        Receiver matcherReceiver;
+        if (receiver instanceof LocalVariableNode) {
+            matcherReceiver = new LocalVariable((LocalVariableNode) receiver);
+        } else {
+            // TODO: what other receiver types should we refine?
+            // Is there a simpler way to get from Node to Receiver?
+            // visitMethodInvocation uses FlowExpressionParseUtil.parse, but
+            // I'm not sure what would be usable to get the receiver of a method call.
+            matcherReceiver = null;
+        }
+
+        // Check matcher is of type java.util.regex.Matcher.
```

Comment:
This could perhaps be factored out into a utility routine that returns a boolean.